include/merlin/utils.cuh

```diff
@@ -299,28 +299,43 @@ struct TypeConvertFunc<int, unsigned int> {
 
 template <class P>
 void realloc(P* ptr, size_t old_size, size_t new_size) {
-  void* new_ptr = nullptr;
+  // Truncate old_size to limit dowstream copy ops.
+  old_size = std::min(old_size, new_size);
```

Comment:
We should not change the old_size in place and it would be better to declare a new var for the min_size.